# Checking Lithium Ion Battery Standardization

Case Illustration:

You work as a data scientist in a lithium ion battery manufacturing company.The production team asks you to check the battery conditions they produce. One of the things requested is to calculate resistance or resistance and battery power every second.

You are asked to do the task as a data scientist in a Lithium Ion battery manufacturing company.These tasks include:

1. Make two explicit functions to calculate the resistance and battery power.Calculate_resistance () function will calculate the resistance value based on the voltage and current given.The Calculate_power () function will calculate the power value based on the voltage and current given.

2. You are asked to create a new column with the name "R/P" inside the dataset.This column will contain the results of the calculation of power/resistance.If the battery capacity value (capacity) is greater or equal to 2.5, the power value will be calculated.However, if the battery capacity value is less than 2.5, the resistance value will be calculated.To do this, you can use the For and IF Statement.You can also use a list to store temporary calculation results before entering it in the "R/P" column.

3. Finally, you are asked to calculate the average battery capacity value (capacity) for data with a current of 1 ampere (current = 1) and on cycle that has an odd number (for example, cycle 1, 3, 5, etc.).

### Imported Library needed

In [1]:
# Add the library usedn
import pandas as pd
import numpy as np

In [2]:
# Enter a dataset that will be used as a test material
data = pd.read_excel("NMC1-9.xlsx")

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90982 entries, 0 to 90981
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   time(s)      90982 non-null  int64  
 1   V_m          90982 non-null  float64
 2   Current (A)  90982 non-null  float64
 3   Cycle        90982 non-null  int64  
 4   Cap          90982 non-null  float64
dtypes: float64(3), int64(2)
memory usage: 3.5 MB


In [4]:
data

time(s)    V_m  Current (A)  Cycle       Cap
0            0  4.098        0.997      1  2.840817
1            1  4.096        0.998      1  2.840817
2            2  4.094        0.999      1  2.840817
3            3  4.093        0.998      1  2.840817
4            4  4.092        0.999      1  2.840817
...        ...    ...          ...    ...       ...
90977    90977  2.506        1.002      9  2.770180
90978    90978  2.504        1.002      9  2.770180
90979    90979  2.502        1.002      9  2.770180
90980    90980  2.500        1.002      9  2.770180
90981    90981  2.497        1.002      9  2.770180

[90982 rows x 5 columns]

Make a calculation function that will be used in testing.
Tests carried out namely to search
1. Obstacles in Ohm Units (R)
    where r = v/i
2. Power in Watt (P)
    where p = v x i i
    
I is the current strength in ampere units

V is a voltage in unit volts

In [5]:
# Function Calculating Resistance
def hitungR(v,i):
    hasil = v/i
    return hasil

# The function of calculating power
def hitungP(v,i):
    hasil = v*i
    return hasil

Next, create a calculation system for test data, where the test conditions carried out are power/resistance with criteria for capacity> = 2.5 then do the power count, capacity < 2.5 then calculate the value of the resistance.

The results of the temporary calculation will be accommodated in the results column and will then be converted to column R/P as a result of the test calculation

In [6]:
hasil = [] # Temporary value column
for i in range(len(data)):
    # Do the condition check on the battery capacity value
    if data["Cap"][i] < 2.5: #Untuk kondisi perhitungan hambatan
        hasil.append(hitungR(data["V_m"][i],data["Current (A)"][i]))
    elif data["Cap"][i] >= 2.5: #Untuk kondisi perhitungan daya
        hasil.append(hitungP(data["V_m"][i],data["Current (A)"][i]))

In [7]:
# Create a new column with the name R/P which is the result of the value of the results column
data["R/P"] = hasil

In [8]:
data

time(s)    V_m  Current (A)  Cycle       Cap       R/P
0            0  4.098        0.997      1  2.840817  4.085706
1            1  4.096        0.998      1  2.840817  4.087808
2            2  4.094        0.999      1  2.840817  4.089906
3            3  4.093        0.998      1  2.840817  4.084814
4            4  4.092        0.999      1  2.840817  4.087908
...        ...    ...          ...    ...       ...       ...
90977    90977  2.506        1.002      9  2.770180  2.511012
90978    90978  2.504        1.002      9  2.770180  2.509008
90979    90979  2.502        1.002      9  2.770180  2.507004
90980    90980  2.500        1.002      9  2.770180  2.505000
90981    90981  2.497        1.002      9  2.770180  2.501994

[90982 rows x 6 columns]

After calculating according to the battery capacity condition and found the R/P column as a result.

Furthermore, calculations are made to find the average value of the R/P value with the condition of the capacity value for current = 1 ampere and odd cycle.

In [9]:
# Displays capacity data (R/P) with the desired conditions
data.loc[(data["Current (A)"] == 1) & (data['Cycle'] % 2 != 0)]

time(s)    V_m  Current (A)  Cycle       Cap    R/P
15          15  4.085          1.0      1  2.840817  4.085
19          19  4.083          1.0      1  2.840817  4.083
20          20  4.083          1.0      1  2.840817  4.083
21          21  4.082          1.0      1  2.840817  4.082
31          31  4.079          1.0      1  2.840817  4.079
...        ...    ...          ...    ...       ...    ...
90909    90909  2.640          1.0      9  2.770180  2.640
90923    90923  2.615          1.0      9  2.770180  2.615
90937    90937  2.589          1.0      9  2.770180  2.589
90949    90949  2.565          1.0      9  2.770180  2.565
90967    90967  2.528          1.0      9  2.770180  2.528

[6830 rows x 6 columns]

In [10]:
# Calculate with the desired conditions for finding the average value
dataBaru = data[(data['Current (A)'] == 1) & (data['Cycle'] % 2 != 0)]
rataRata = np.mean(dataBaru['Cap'])

print("The average value of the capacity value for current = 1 ampere and odd cycle is", rataRata)

The average value of the capacity value for current = 1 ampere and odd cycle is 2.8064845553927484


### Results

After the experiment, then in checking in the standardization of lithium ion batteries found the results of testing in the Cap column.

In addition, also found the average value of battery capacity with the required conditions found an average value of 2.8064845553927484